scripts/usb_reset.py

```
@@ -14,46 +14,266 @@
+                        help="reset device only, for privileged mode")
+
+    detach = subparsers.add_parser("detach", help="detach a USB device")
+    detach.add_argument("device", help="the target usb device")
+    detach.add_argument("-d", dest="dom_id", type=int, required=True,
+                        help="specify the dom_id of the VM")
+    detach.add_argument("-i", dest="id", type=int, nargs=2, required=True,
+                        help="restore ownership to uid gid")
+
+    cleanup = subparsers.add_parser("cleanup", help="clean up chroot directory")
+    cleanup.add_argument("-d", dest="dom_id", type=int, required=True,
+                         help="specify the dom_id of the VM")
+
+    return parser.parse_args()
 
-#USBDEVFS_RESET _IO('U', 20)
-USBDEVFS_RESET = (ord('U') << 8) | 20
 
+def get_root_dir(dom_id):
+    return "/var/xen/qemu/{}/".format(dom_id)
 
+
+def get_cg_dir(dom_id):
+    return "/sys/fs/cgroup/devices/qemu-{}/".format(dom_id)
+
+
+# throw IOError, ValueError
 def read_int(path):
     with open(path) as f:
         return int(f.readline())
 
 
-def reset_device(bus, dev):
-    with open("/dev/bus/usb/{0:03d}/{1:03d}".format(bus, dev), "w") as f:
-        fcntl.ioctl(f.fileno(), USBDEVFS_RESET, 0)
+def dev_path(device):
+    pat = re.compile(r"^\d+-\d+(\.\d+)*$")
+    if pat.match(device) is None:
+        log.debug("Unexpected device node: {}".format(device))
+        exit(1)
+    try:
+        bus = read_int("/sys/bus/usb/devices/{}/busnum".format(device))
+        dev = read_int("/sys/bus/usb/devices/{}/devnum".format(device))
+        return "/dev/bus/usb/{0:03d}/{1:03d}".format(bus, dev)
+    except (IOError, ValueError) as e:
+        log.debug("Failed to get device path {}: {}".format(device, str(e)))
+        exit(1)
 
 
-if __name__ == "__main__":
-    log.logToSyslog(level=logging.DEBUG)
+def get_ctl(path, mode):
+    try:
+        st = os.stat(path)
+    except OSError:
+        return None
```

Why are we ignoring errors? I think any error setting up a device should be fatal.

---

done.

---

can I log and exit(1)?

---

You could, but it would be better practice to log and raise an exception.

---

I don't like to re-throw the exception, because we don't handle it somewhere(level) else. Any error is fatal.

---

It looks to me that both callers, `_device_ctl()` and `setup_cgroup()` catch the exception and exit appropriately.

---

If so, do we have to raise another run-time exception at the end of get_ctrl, instead of exit(1)?

---

Yes, that's probably better.